src/primitive_esp32.cc

```diff
@@ -72,6 +72,184 @@ namespace toit {
+    }
+  }
+
+  // The last OTA is the only one that is allowed to not be divisible
+  // by 16.
+  if (ota_written != Utils::round_up(ota_written, FLASH_SEGMENT_SIZE)) {
+    ESP_LOGE("Toit", "More OTA was written after last block");
+    OUT_OF_BOUNDS;
+  }
+
+  if (ota_size > 0 && (ota_written + bytes.length() > ota_size)) {
+    ESP_LOGE("Toit", "OTA write overflows predetermined size (%d + %d > %d)",
+        ota_written, bytes.length(), ota_size);
+    OUT_OF_BOUNDS;
+  }
+
+  uword to_write = Utils::round_down(bytes.length(), FLASH_SEGMENT_SIZE);
+
+  word erase_from = Utils::round_up(ota_written, FLASH_PAGE_SIZE);
+  word erase_to = Utils::round_up(ota_written + to_write, FLASH_PAGE_SIZE);
+  for (word page = erase_from; page < erase_to; page += FLASH_PAGE_SIZE) {
+    esp_err_t err = esp_partition_erase_range(ota_partition, page, FLASH_PAGE_SIZE);
+    if (err != ESP_OK) {
+      ota_partition = null;
+      ESP_LOGE("Toit", "esp_partition_erase_range failed (%s)", esp_err_to_name(err));
+      OUT_OF_BOUNDS;
+    }
+  }
+
+  esp_err_t err = esp_partition_write(ota_partition, ota_written, reinterpret_cast<const void*>(bytes.address()), to_write);
+
+  if (err == ESP_OK && to_write != bytes.length()) {
+    // Last write can be a non-multiple of 16.  We pad it up.
+    uint8 temp_buffer[FLASH_SEGMENT_SIZE];
+    memset(temp_buffer, 0, FLASH_SEGMENT_SIZE);
+    memcpy(temp_buffer, bytes.address() + to_write, bytes.length() - to_write);
+    err = esp_partition_write(ota_partition, ota_written, reinterpret_cast<const void*>(temp_buffer), bytes.length() - to_write);
+  }
+
+  if (err != ESP_OK) {
+    ESP_LOGE("Toit", "esp_partition_write failed (%s)!", esp_err_to_name(err));
+    ota_partition = null;
+    OUT_OF_BOUNDS;
+  }
+
+  ota_written += bytes.length();
+  return Smi::from(ota_written);
+}
+
+PRIMITIVE(ota_end) {
+  PRIVILEGED;
+  ARGS(int, size, Object, expected);
+  esp_err_t err = ESP_OK;
+
+  if (size != 0) {
+    if (ota_partition == null) {
+      ESP_LOGE("Toit", "Cannot end OTA session before starting it");
+      OUT_OF_BOUNDS;
+    }
+
+    ASSERT(ota_size == 0 || (ota_written <= ota_size));
+    if (ota_size > 0 && ota_written < ota_size) {
+      ESP_LOGE("Toit", "OTA only partially written (%d < %d)", ota_written, ota_size);
+      OUT_OF_BOUNDS;
+    }
+
+    const esp_partition_pos_t partition_position = {
+      ota_partition->address,
+      ota_partition->size
+    };
+
+    esp_image_metadata_t image_metadata;
+
+    err = esp_image_verify(ESP_IMAGE_VERIFY, &partition_position, &image_metadata);
+    if (err != ESP_OK) {
+      ESP_LOGE("Toit", "esp_image_verify failed (%s)!", esp_err_to_name(err));
+      ota_partition = null;
+      OUT_OF_BOUNDS;
+    }
+
+    // The system SHA256 checksum is optional, so we add an explicit verification
+    // that we control.  (There is also a non-optional checksum, but it is only one
+    // byte, and so not really reliable.)
+    if (expected->is_byte_array()) {
```

Comment:
This should use expected->byte_content so that it works with slices too and CoW-byte arrays too.